### PR TITLE
fix: precision loss and inaccurate limit prices when confirming limit order

### LINKED ISSUE #8317: IMPORTANT:  Override CoW quote amounts before submitting limit order
### Overview

Limit prices between the input and the preview differ because the preview uses the actual quoted value while the input uses the inputted values:

This is likely (TBC) because the quote is for immediate spot execution whereas a limit order has a fixed price, so the current implementation is WRONG and should be fixed before enabling limit orders in production.

### References and additional details

Cow recommends overwriting the quoted amounts when submitting an order:
https://learn.cow.fi/tutorial/submit-order
```typescript
const orderId = await orderBookApi.sendOrder({
  ...quote,
  ...orderSigningResult,
  sellAmount: order.sellAmount, // replace quote sellAmount with signed order sellAmount, which is equal to original sellAmount
  feeAmount: order.feeAmount, // replace quote feeAmount with signed order feeAmount, which is 0
  signingScheme: orderSigningResult.signingScheme as unknown as SigningScheme
})
```

Context https://github.com/shapeshift/web/pull/8299#issuecomment-2527316668

### Acceptance Criteria

- Submitted limit orders match user input, NOT the spot quoted values.

### Need By Date

_No response_

### Screenshots/Mockups

_No response_

### Estimated effort

_No response_

### PR DESCRIPTION
**Pull request by woodenfurniture**
## Description

The limit price and amounts displayed in the limit order confirmation UI don't match user inputs (which is bad). This PR addresses this by:
- centralizing calculation logic
- ensuring the same calculations are done between the input and confirmation UIs
- ignoring the quoted sell amount and using the inputted sell amount (per the [cowswap docs](https://learn.cow.fi/tutorial/submit-order))

Includes some scary code comments to try and stop us breaking it in the future.

## Issue (if applicable)

closes #8317

## Risk

> High Risk PRs Require 2 approvals

Low risk as this is not yet user facing and (should) only affect display logic. The amounts being submitted to cowswap should be unaffected (which likely were correct). Still, it would be wise to review with caution as getting this wrong would be disastrous for users when this goes live.

> What protocols, transaction types, wallets or contract interactions might be affected by this PR?

## Testing

Check the amounts displayed in the limit order confirmation UI matches the inputted values.

### Operations

- [ ] :checkered_flag: My feature is behind a flag and doesn't require operations testing (yet)

## Screenshots (if applicable)

Demo showing amounts matching. See #8317 for screenshots of original issue.
https://jam.dev/c/62a219fc-dded-4c21-a122-22e1a9230f08

### CODE REVIEW

File: src/components/MultiHopTrade/components/LimitOrder/components/LimitOrderInput.tsx [hunk 1]
```diff
@@ -234,10 +235,13 @@ export const LimitOrderInput = ({
...
-      .div(fromBaseUnit(quoteResponse.quote.sellAmount, sellAsset.precision))
-      .toFixed()
-  }, [buyAsset.precision, quoteResponse, sellAsset.precision, limitOrderQuoteParams])
+    return calcLimitPriceBuyAsset({
+      sellAmountCryptoBaseUnit: quoteResponse.quote.sellAmount,
+      buyAmountCryptoBaseUnit: quoteResponse.quote.buyAmount,
+      sellAsset,
+      buyAsset,
+    })
+  }, [quoteResponse, limitOrderQuoteParams, sellAsset, buyAsset])
```
Comment (gomesalexandre) [inline src/components/MultiHopTrade/components/LimitOrder/components/LimitOrderInput.tsx:244]: 💜

File: src/state/slices/limitOrderSlice/helpers.ts [hunk 2]
```diff
@@ -9,6 +10,9 @@ export const buildUnsignedLimitOrder = ({
 }: LimitOrderActiveQuote): Omit<OrderCreation, 'signature'> => {
   return {
     ...response.quote,
+    // IMPORTANT: Replace the amounts with the user inputted values. We don't want to use the quoted
+    // amounts here because the quote amounts assume spot trade and will not match the user-inputted
+    // limit price.
```
Comment (gomesalexandre) [inline src/state/slices/limitOrderSlice/helpers.ts:15]: 💜

File: src/state/slices/limitOrderSlice/selectors.ts [hunk 3]
```diff
@@ -41,25 +42,47 @@ export const selectActiveQuoteFeeAsset = (state: ReduxState) => {
...
+)
+
+export const selectActiveQuoteBuyAmountCryptoBaseUnit = createSelector(
   selectActiveQuote,
+  activeQuote => {
+    if (!activeQuote) return '0'
+
+    // DANGER: DO NOT use the values in the quote response - the quote amounts assume spot trade and
+    // will not match the user-inputted limit price. Doing so will discard the user input limit and
+    // do a spot trade.
```
Comment (gomesalexandre) [inline src/state/slices/limitOrderSlice/selectors.ts:64]: 💜

File: src/state/slices/limitOrderInputSlice/selectors.ts [hunk 4]
```diff
@@ -63,14 +63,10 @@ export const selectExpiry = createSelector(selectBaseSlice, baseSlice => baseSli
...
   selectLimitPrice,
-  (inputSellAmountCryptoPrecision, buyAsset, limitPrice) => {
-    // Arithmetic MUST be in CryptoPrecision due to differing decimals on various tokens.
-    return toBaseUnit(
-      bn(inputSellAmountCryptoPrecision).times(limitPrice.buyAssetDenomination),
-      buyAsset.precision,
-    )
+  (inputSellAmountCryptoBaseUnit, limitPrice) => {
+    // Arithmetic MUST be in CryptoBaseUnit to avoid precision loss converting back and forth.
+    return bn(inputSellAmountCryptoBaseUnit).times(limitPrice.buyAssetDenomination).toFixed()
```
Comment (gomesalexandre) [inline src/state/slices/limitOrderInputSlice/selectors.ts:70]: blocking: this won't work when dealing with multiple base units e.g:

image

- 100123456781234567000 inputSellAmountCryptoBaseUnit (100.123456781234567000 FOX)
- 0.07238467339191986179 `limitPrice.buyAssetDenomination`

Yields `7247403717979668012.23008076421049493` which is off by sellAssetPrecision - buyAssetPrecision i.e 18 - 6 = 18

**Review by gomesalexandre — CHANGES_REQUESTED** [review]
Tested locally, confirmed this fixes the sell amounts, however brings regression in buy amounts/rates, see https://github.com/shapeshift/web/pull/8323/files#r1877396425

- Sell amount is now matching user input ✅ 

See against develop (right)

image
image
image

- USDC -> FOX: FOX limit price is zero'd out, as well as the amount of received FOX 🚫 

Confirmed this is a regression against develop (right)

image

- 100 FOX -> USDC shows erroneous values (off by 10^12 i.e 7ish USDC expected vs. 7ish Trillion actual) 🚫 

image

And similarly for the limit price being off by 10^12

File: src/state/slices/limitOrderInputSlice/selectors.ts (hunk 4 above)
Comment (woodenfurniture) [reply src/state/slices/limitOrderInputSlice/selectors.ts]: Actually really well spotted. I've pushed a fix 3da61fbf5ed5eaa0058c36f4f2192304e559db1b to convert the precision to the buy asset precision here, and it looks to have resolved the issues noted here.

**Review by gomesalexandre — APPROVED** [review]
Re-tested locally against develop (right), confirmed second time's the charm! This still fixes the sell amounts, and regression in buy amounts/rates is now gone 🎉 

- Sell amount is now matching user input (market) ✅ 

image
image

image
image

- Sell amount is now matching user input (percent option) ✅ 

image
image

image
image

- Sell amount is now matching user input (manual user input) ✅ 

image
image

image
image


- USDC -> FOX: FOX receive value is correct ✅ 

image
image

- FOX->USDC: USDC receive value is correct ✅ 

image

**Pull request merged**